sql/src/main/java/io/crate/jobs/JobExecutionContext.java

```diff
@@ -205,89 +216,74 @@ public void close() {
                 }
             }
         }
+        try {
+            chainedFuture.get();
+        } catch (Exception e) {
+            Throwables.propagate(e);
+        }
     }
 
     private void callCloseCallback() {
         if (closeCallback == null) {
             return;
         }
-        if (activeSubContexts.get() == 0) {
-            closeCallback.handle(this);
-        }
+        closeCallback.handle(this);
     }
 
     @Override
     public String toString() {
         return "JobExecutionContext{" +
                 "jobId=" + jobId +
-                ", subContexts=" + subContexts +
-                ", activeSubContexts=" + activeSubContexts +
+                ", activeSubContexts=" + subContexts.size() +
                 ", closed=" + closed +
                 '}';
     }
 
-    private class RemoveContextCallback implements ContextCallback {
+    private class RemoveSubContextCallback implements FutureCallback<SubExecutionContextFuture.State> {
 
-        private final int executionPhaseId;
+        private final int id;
 
-        public RemoveContextCallback(int executionPhaseId) {
-            this.executionPhaseId = executionPhaseId;
+        private RemoveSubContextCallback(int id) {
+            this.id = id;
         }
 
-        private void close(@Nullable Throwable error, long bytesUsed) {
-            Object remove;
-            synchronized (subContexts) {
-                remove = subContexts.remove(executionPhaseId);
-            }
+        private boolean remove(){
```

Comment:
consider using an enum instead of boolean as return value for better readability